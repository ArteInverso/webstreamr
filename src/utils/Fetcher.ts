import CachePolicy from 'http-cache-semantics';
import TTLCache from '@isaacs/ttlcache';
import winston from 'winston';
import { Context } from '../types';
import { BlockedError, NotFoundError } from '../error';
import { clearTimeout } from 'node:timers';

interface HttpCacheItem { policy: CachePolicy; status: number; statusText: string; body: string }

export class Fetcher {
  private readonly logger: winston.Logger;
  private readonly httpCache: TTLCache<string, HttpCacheItem>;

  constructor(logger: winston.Logger) {
    this.logger = logger;
    this.httpCache = new TTLCache();
  }

  readonly text = async (ctx: Context, url: URL, init?: RequestInit): Promise<string> => {
    return (await this.cachedFetch(ctx, url, init)).body;
  };

  readonly textPost = async (ctx: Context, url: URL, data: unknown, init?: RequestInit): Promise<string> => {
    return (await this.cachedFetch(ctx, url, { ...init, method: 'POST', body: JSON.stringify(data) })).body;
  };

  readonly head = async (ctx: Context, url: URL, init?: RequestInit): Promise<CachePolicy.Headers> => {
    return (await this.cachedFetch(ctx, url, { ...init, method: 'HEAD' })).policy.responseHeaders();
  };

  private readonly getInit = (ctx: Context, url: URL, init?: RequestInit): RequestInit => ({
    ...init,
    headers: {
      'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
      'Accept-Language': 'en',
      'Forwarded': `for=${ctx.ip}`,
      'Priority': 'u=0',
      'Referer': `${ctx.referer?.href ?? url.origin}`,
      'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/134.0.0.0 Safari/537.3',
      'X-Forwarded-For': ctx.ip,
      'X-Forwarded-Proto': url.protocol.slice(0, -1),
      'X-Real-IP': ctx.ip,
      ...init?.headers,
    },
  });

  private readonly handleHttpCacheItem = (httpCacheItem: HttpCacheItem): HttpCacheItem => {
    if (httpCacheItem.status && httpCacheItem.status >= 200 && httpCacheItem.status <= 299) {
      return httpCacheItem;
    }

    if (httpCacheItem.status === 404) {
      throw new NotFoundError();
    }

    const responseHeaders = httpCacheItem.policy.responseHeaders();

    if (httpCacheItem.policy.responseHeaders()['cf-mitigated'] === 'challenge') {
      throw new BlockedError('cloudflare_challenge');
    }

    if (httpCacheItem.status === 403) {
      throw new BlockedError('unknown');
    }

    throw new Error(`Fetcher error: ${httpCacheItem.status}: ${httpCacheItem.statusText}, response headers: ${JSON.stringify(responseHeaders)}`);
  };

  private readonly headersToObject = (headers: Headers): Record<string, string> => {
    const obj: Record<string, string> = {};

    headers.forEach((value, name) => {
      obj[name] = value;
    });

    return obj;
  };

  private readonly cachedFetch = async (ctx: Context, url: URL, init?: RequestInit): Promise<HttpCacheItem> => {
    const newInit = this.getInit(ctx, url, init);

    const request: CachePolicy.Request = { url: url.href, method: newInit.method ?? 'GET', headers: {} };

    let httpCacheItem: HttpCacheItem | undefined = this.httpCache.get(url.href);
    if (httpCacheItem) {
      this.logger.info(`Cached fetch ${request.method} ${url}`, ctx);
      return this.handleHttpCacheItem(httpCacheItem);
    }

    this.logger.info(`Fetch ${request.method} ${url}`, ctx);

    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), 10000);

    let response;
    try {
      response = await fetch(url, { ...newInit, signal: controller.signal });
    } finally {
      clearTimeout(timer);
    }

    const body = await response.text();

    const policy = new CachePolicy(request, { status: response.status, headers: this.headersToObject(response.headers) }, { shared: false });

    httpCacheItem = { policy, status: response.status, statusText: response.statusText, body };

    let ttl;
    if (response.ok) {
      ttl = Math.max(policy.timeToLive(), 900000); // 15m at least
    } else {
      ttl = policy.timeToLive();
    }

    if (ttl > 0) {
      this.httpCache.set(url.href, httpCacheItem, { ttl });
    }

    return this.handleHttpCacheItem(httpCacheItem);
  };
}
