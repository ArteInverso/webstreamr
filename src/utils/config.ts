import { Config } from '../types';

export const getDefaultConfig = (): Config => {
  return { multi: 'on', en: 'on' };
};

export const showExternalUrls = (config: Config): boolean => 'includeExternalUrls' in config;

export const hasMultiEnabled = (config: Config): boolean => 'multi' in config;
