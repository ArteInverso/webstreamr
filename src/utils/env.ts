export const envGet = (name: string): string | undefined => process.env[name];

export const envGetAppId = (): string => process.env['MANIFEST_ID'] || 'webstreamr';

export const envGetAppName = (): string => process.env['MANIFEST_NAME'] || 'WebStreamr';

export const envIsProd = (): boolean => process.env['NODE_ENV'] === 'production';
