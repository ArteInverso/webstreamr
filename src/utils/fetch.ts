import makeFetchHappen from 'make-fetch-happen';
import { FetchInterface, FetchOptions } from 'make-fetch-happen';
import fs from 'node:fs';
import * as os from 'node:os';
import { Response } from 'node-fetch';
import { logInfo } from './log';

// eslint-disable-next-line @typescript-eslint/no-extraneous-class
class MakeFetchHappenSingleton {
  private static instance: FetchInterface;

  public static getInstance(): FetchInterface {
    if (!MakeFetchHappenSingleton.instance) {
      MakeFetchHappenSingleton.instance = makeFetchHappen.defaults({
        cachePath: `${fs.realpathSync(os.tmpdir())}/webstreamr`,
      });
    }
    return MakeFetchHappenSingleton.instance;
  }
}

const cachedFetch = async (uriOrRequest: string | Request, opts?: FetchOptions): Promise<Response> => {
  logInfo(`Fetch ${uriOrRequest}`);

  const fetch = MakeFetchHappenSingleton.getInstance();

  return await fetch(uriOrRequest, opts);
};

export const cachedFetchText = async (uriOrRequest: string | Request, opts?: FetchOptions): Promise<string> => {
  const response = await cachedFetch(uriOrRequest, opts);
  if (!response.ok) {
    throw new Error(`HTTP error: ${response.status} - ${response.statusText}`);
  }

  return await response.text();
};
