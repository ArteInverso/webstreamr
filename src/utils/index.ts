export * from './env';
export * from './Fetcher';
export * from './StreamResolver';
export * from './embed';
export * from './imdb';
export * from './languageFromCountryCode';
export * from './manifest';
export * from './tmdb';
