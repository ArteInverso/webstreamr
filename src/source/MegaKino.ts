import * as cheerio from 'cheerio';
import { ContentType } from 'stremio-addon-sdk';
import { Context, CountryCode } from '../types';
import { Fetcher, getImdbId, Id, ImdbId } from '../utils';
import { Source, SourceResult } from './types';

export class MegaKino implements Source {
  public readonly id = 'megakino';

  public readonly label = 'MegaKino';

  public readonly contentTypes: ContentType[] = ['movie'];

  public readonly countryCodes: CountryCode[] = [CountryCode.de];

  private readonly fetcher: Fetcher;

  public constructor(fetcher: Fetcher) {
    this.fetcher = fetcher;
  }

  public async handle(ctx: Context, _type: string, id: Id): Promise<SourceResult[]> {
    const imdbId = await getImdbId(ctx, this.fetcher, id);

    const pageUrl = await this.fetchPageUrl(ctx, imdbId);
    if (!pageUrl) {
      return [];
    }

    const html = await this.fetcher.text(ctx, pageUrl);

    const $ = cheerio.load(html);

    const title = $('title').text().trim();

    return Promise.all(
      $(`.video-inside iframe`)
        .map((_i, el) => new URL(($(el).attr('data-src') ?? $(el).attr('src')) as string))
        .toArray()
        .map(url => ({ countryCode: CountryCode.de, referer: pageUrl, title, url })),
    );
  };

  private fetchPageUrl = async (ctx: Context, imdbId: ImdbId): Promise<URL | undefined> => {
    const html = await this.fetcher.text(ctx, new URL(`https://megakino.si/?do=search&subaction=search&story=${imdbId.id}`));

    const $ = cheerio.load(html);

    const url = $('#dle-content a[href].poster:first')
      .map((_i, el) => $(el).attr('href'))
      .get(0);

    return url !== undefined ? new URL(url) : url;
  };
}
