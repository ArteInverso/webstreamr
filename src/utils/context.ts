import { Request } from 'express';
import { Context } from '../types';
import { getDefaultConfig } from './config';

export const contextFromRequest = (req: Request): Context => {
  return {
    hostUrl: new URL(`${req.protocol}://${req.host}`),
    id: req.headers['X-Request-ID'] as string,
    ...(req.ip && { ip: req.ip }),
    config: req.params['config'] ? JSON.parse(req.params['config']) : getDefaultConfig(),
  };
};
