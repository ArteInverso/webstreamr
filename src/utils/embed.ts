import { unpack } from 'unpacker';
import { cachedFetchText } from './fetch';
import { scanFromResolution } from './resolution';

const LINK_REGEXPS = [
  /sources:\[{file:"(.*?)"/, // dropload, supervideo
];

interface ParsePackedEmbedResult {
  url: string;
  resolution: string | undefined;
  size: string | undefined;
}

const findResolution = (html: string): string | undefined => {
  const match = html.match(/(\d{3,}x\d{3,}),/);

  return match && match[1] ? scanFromResolution(match[1]) : undefined;
};

const findSize = (html: string): string | undefined => {
  const sizeMatch = html.match(/([\d.]+) ?([GM]B)/);

  return sizeMatch && sizeMatch[1] && sizeMatch[2] ? `${sizeMatch[1]} ${sizeMatch[2]}` : undefined;
};

export const parsePackedEmbed = async (url: string): Promise<ParsePackedEmbedResult> => {
  const html = await cachedFetchText(url);

  const evalMatch = html.match(/eval\(function\(p,a,c,k,e,d\).*\)\)/);
  if (!evalMatch) {
    throw new Error(`No p.a.c.k.e.d string found on ${url}`);
  }

  const unpacked = unpack(evalMatch[0]);

  for (const linkRegexp of LINK_REGEXPS) {
    const linkMatch = unpacked.match(linkRegexp);
    if (linkMatch && linkMatch[1]) {
      return {
        url: 'https://' + linkMatch[1].replace(/^(https:)?\/\//, ''),
        resolution: findResolution(html),
        size: findSize(html),
      };
    }
  }

  throw new Error(`Could not find a stream link on ${url}`);
};
