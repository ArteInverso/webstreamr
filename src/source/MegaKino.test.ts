import { createTestContext } from '../test';
import { FetcherMock, ImdbId } from '../utils';
import { MegaKino } from './MegaKino';

const ctx = createTestContext({ de: 'on' });

describe('MegaKino', () => {
  let handler: MegaKino;

  beforeEach(() => {
    handler = new MegaKino(new FetcherMock(`${__dirname}/__fixtures__/MegaKino`));
  });

  test('handles non-existent movies gracefully', async () => {
    const streams = await handler.handle(ctx, 'movie', new ImdbId('tt12345678', undefined, undefined));
    expect(streams).toHaveLength(0);
  });

  test('handle imdb baymax', async () => {
    const streams = await handler.handle(ctx, 'movie', new ImdbId('tt2245084', undefined, undefined));
    expect(streams).toMatchSnapshot();
  });
});
