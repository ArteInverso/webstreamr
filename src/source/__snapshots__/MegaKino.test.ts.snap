// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`MegaKino handle imdb baymax 1`] = `
[
  {
    "countryCode": "de",
    "referer": "https://megakino.si/films/693-baymax-riesiges-robowabohu.html",
    "title": "Baymax - Riesiges Robowabohu",
    "url": "https://voe.sx/e/8doq7gwiwtig",
  },
  {
    "countryCode": "de",
    "referer": "https://megakino.si/films/693-baymax-riesiges-robowabohu.html",
    "title": "Baymax - Riesiges Robowabohu",
    "url": "https://watch.gxplayer.xyz/watch?v=MEKI92PU",
  },
]
`;
